manifests/config/server/tomcat_users.pp

```diff
@@ -0,0 +1,97 @@
+# Definition tomcat::config::server::tomcat_users
+#
+# Configures roles and users in $CATALINA_BASE/conf/tomcat-users.xml
+# or any other specified file
+#
+# Parameters:
+# - $catalina_base is the base directory for the Tomcat installation
+# - $element specifies the element type. Valid values are 'user' or 'role'.
+#   Defaults to 'user'.
+# - $element_name sets the 'username' or 'rolename'. Depends on the $element.
+#   Defaults to $name.
+# - $ensure specifies whether you are trying to add or remove the element.
+#   Valid values are 'present', 'absent', 'true', and 'false'. Defaults to
+#   'present'.
+# - $file: The path to the file to manage. Must be fully qualified.
+#   Defaults to $CATALINA_BASE/conf/tomcat-users.xml.
+# - Set $manage_file to true for managing the file. It sets file permission,
+#   owner, group and create a basic tomcat-users XML if file does not exist yet.
+# - $password specifies the password for a user ($element = 'user').
+# - $roles specifies the roles for a user ($element = 'user').
+#
+define tomcat::config::server::tomcat_users (
+  $catalina_base = $::tomcat::catalina_home,
+  $element       = 'user',
+  $element_name  = undef,
+  $ensure        = present,
+  $file          = undef,
+  $manage_file   = true,
+  $password      = undef,
+  $roles         = [],
+) {
+
+  if versioncmp($::augeasversion, '1.0.0') < 0 {
+    fail('Server configurations require Augeas >= 1.0.0')
+  }
+
+  validate_re($element, '^(user|role)$')
+  validate_re($ensure, '^(present|absent|true|false)$')
+  validate_array($roles)
+  validate_bool($manage_file)
+
+  if $element == 'role' and ( $password or ! empty($roles) ) {
+    warning('$password and $roles are useless when $element is set to \'role\'')
+  }
+  
+  if $element == 'user' {
+    $element_identifier = 'username'
+  } else {
+    $element_identifier = 'rolename'
+  }
+
+  if $element_name {
+    $_element_name = $element_name
+  } else {
+    $_element_name = $name
+  }
+  
+  if $file {
+    $_file = $file
+  } else {
+    $_file = "${catalina_base}/conf/tomcat-users.xml"
+  }
+  
+  if $manage_file {
+    ensure_resource('file', $_file, {
+      ensure  => file,
+      path    => $_file,
+      replace => false,
+      content => '<?xml version=\'1.0\' encoding=\'utf-8\'?><tomcat-users></tomcat-users>',
+      owner   => $::tomcat::user,
+      group   => $::tomcat::group,
+      mode    => '0640',
+    })
+  }

```

Comment:
This block is going to cause issues if configuration is actually being managed with augeas. On every puppet run the file content will change. I'm not sure what the best way to create the file if it doesn't exist, as much as I don't like it an exec with an onlyif to make sure the file doesn't already exist might be the best way to go. Or we might be able to do something with augeas to create the base elements.

Whatever ends up happening to this resource we also want to make sure it is executed before the other augeas resource below.